File: internal/provenance/provenance.go
package provenance

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// Record is embedded in every run artifact for paper-grade reproducibility.
type Record struct {
	HarnessCommit string `json:"harness_commit"`
	HarnessDirty  bool   `json:"harness_dirty"`
	GoVersion     string `json:"go_version,omitempty"`
	JavaVersion   string `json:"java_version,omitempty"`
	PythonVersion string `json:"python_version,omitempty"`
	ProfilePath   string `json:"profile_path"`
	ProfileSHA256 string `json:"profile_sha256"`
	SuiteID       string `json:"suite_id,omitempty"`
	SuiteSeed     int    `json:"suite_seed,omitempty"`
	SuiteN        int    `json:"suite_n,omitempty"`
	RunIndex      int    `json:"run_index"`
	Driver        string `json:"driver"`
	Variant       string `json:"variant"`
	Middleware    string `json:"middleware,omitempty"`
	CapturedAtUTC string `json:"captured_at_utc"`
}

// Options configures provenance collection for a single run.
type Options struct {
	ProfilePath string
	SuiteID     string
	SuiteSeed   int
	SuiteN      int
	RunIndex    int
	Driver      string
	Variant     string
	Middleware  string
	JavaVersion string
	PythonVer   string
}

// Collect builds a provenance record from the environment and profile file.
func Collect(opts Options) (Record, error) {
	commit, dirty := resolveHarnessCommit()

	profSHA, err := fileSHA256(opts.ProfilePath)
	if err != nil {
		return Record{}, fmt.Errorf("profile hash: %w", err)
	}

	runIndex := opts.RunIndex
	if runIndex < 1 {
		runIndex = 1
	}

	return Record{
		HarnessCommit: commit,
		HarnessDirty:  dirty,
		GoVersion:     runtime.Version(),
		JavaVersion:   opts.JavaVersion,
		PythonVersion: opts.PythonVer,
		ProfilePath:   repoRelativePath(opts.ProfilePath),
		ProfileSHA256: profSHA,
		SuiteID:       opts.SuiteID,
		SuiteSeed:     opts.SuiteSeed,
		SuiteN:        opts.SuiteN,
		RunIndex:      runIndex,
		Driver:        opts.Driver,
		Variant:       opts.Variant,
		Middleware:    opts.Middleware,
		CapturedAtUTC: time.Now().UTC().Format(time.RFC3339),
	}, nil
}

// repoRelativePath stores profile_path relative to the harness module root when possible.
func repoRelativePath(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return filepath.ToSlash(path)
	}
	root, err := findModuleRoot()
	if err != nil {
		return filepath.ToSlash(abs)
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil {
		return filepath.ToSlash(abs)
	}
	return filepath.ToSlash(rel)
}

func resolveHarnessCommit() (string, bool) {
	if env := strings.TrimSpace(os.Getenv("EMRTD_HARNESS_COMMIT")); env != "" {
		return env, false
	}
	commit, dirty, err := gitHead()
	if err != nil || commit == "" {
		return "unknown", false
	}
	return commit, dirty
}

func gitHead() (string, bool, error) {
	root, err := findModuleRoot()
	if err != nil {
		return "", false, err
	}
	out, err := exec.Command("git", "-C", root, "rev-parse", "HEAD").Output()
	if err != nil {
		return "", false, err
	}
	commit := strings.TrimSpace(string(out))
	if commit == "" {
		return "", false, fmt.Errorf("empty git rev-parse HEAD")
	}
	status, _ := exec.Command("git", "-C", root, "status", "--porcelain").Output()
	return commit, len(strings.TrimSpace(string(status))) > 0, nil
}

func findModuleRoot() (string, error) {
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir, nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", fmt.Errorf("go.mod not found")
		}
		dir = parent
	}
}

func fileSHA256(path string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(raw)
	return hex.EncodeToString(sum[:]), nil
}

File: classifier/provenance.py
"""Experiment provenance helpers (shared with internal/provenance)."""

from __future__ import annotations

import hashlib
import json
import os
import platform
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def _git_head(root: Path) -> tuple[str, bool]:
    env = (os.environ.get("EMRTD_HARNESS_COMMIT") or "").strip()
    if env:
        return env, False
    try:
        commit = subprocess.check_output(
            ["git", "-C", str(root), "rev-parse", "HEAD"], text=True
        ).strip()
        if not commit:
            return "unknown", False
        status = subprocess.check_output(
            ["git", "-C", str(root), "status", "--porcelain"], text=True
        ).strip()
        return commit, bool(status)
    except (subprocess.CalledProcessError, FileNotFoundError):
        return "unknown", False


def file_sha256(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def _captured_at_utc() -> str:
    raw = (os.environ.get("SOURCE_DATE_EPOCH") or "").strip()
    if raw:
        try:
            return datetime.fromtimestamp(int(raw), tz=timezone.utc).strftime(
                "%Y-%m-%dT%H:%M:%SZ"
            )
        except ValueError:
            pass
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _gmrtd_commit(root: Path) -> str | None:
    env = (os.environ.get("GMRTD_COMMIT") or "").strip()
    if env:
        return env
    for candidate in (
        (os.environ.get("GMRTD_PATH") or "").strip(),
        str(root.parent / "_vendor" / "gmrtd"),
        str(root / "_vendor" / "gmrtd"),
    ):
        if not candidate:
            continue
        try:
            return subprocess.check_output(
                ["git", "-C", candidate, "rev-parse", "HEAD"], text=True
            ).strip()
        except (subprocess.CalledProcessError, FileNotFoundError):
            continue
    return None


def collect(
    *,
    root: Path,
    profile_path: Path,
    suite_id: str,
    suite_seed: int,
    suite_n: int,
    run_index: int,
    driver: str,
    variant: str,
    middleware: str = "",
    java_version: str = "",
) -> dict[str, Any]:
    commit, dirty = _git_head(root)
    if run_index < 1:
        run_index = 1
    try:
        rel = profile_path.resolve().relative_to(root.resolve())
        profile_stored = rel.as_posix()
    except ValueError:
        profile_stored = profile_path.as_posix()
        marker = "/emrtd-differential-harness/"
        if marker in profile_stored:
            profile_stored = profile_stored.split(marker, 1)[1]
    return {
        "harness_commit": commit or "unknown",
        "harness_dirty": dirty,
        "gmrtd_commit": _gmrtd_commit(root),
        "python_version": platform.python_version(),
        "java_version": java_version or None,
        "profile_path": profile_stored,
        "profile_sha256": file_sha256(profile_path),
        "suite_id": suite_id,
        "suite_seed": suite_seed,
        "suite_n": suite_n,
        "run_index": run_index,
        "driver": driver,
        "variant": variant,
        "middleware": middleware or None,
        "captured_at_utc": _captured_at_utc(),
    }


def tool_versions(root: Path) -> dict[str, str]:
    versions: dict[str, str] = {"python": platform.python_version()}
    for name, cmd in (
        ("go", ["go", "version"]),
        ("java", ["java", "-version"]),
    ):
        try:
            proc = subprocess.run(cmd, capture_output=True, text=True, check=False)
            versions[name] = (proc.stdout or proc.stderr).strip().splitlines()[0]
        except FileNotFoundError:
            versions[name] = "not found"
    return versions
